pdudaemon/drivers/conrad197720.py

```diff
@@ -0,0 +1,354 @@
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+# Script to control conrad components relais card 197720
+# Up to 255 cards can be concatenated at one serial port
+# https://www.conrad.de/de/p/conrad-components-197720-relaiskarte-baustein-12-v-dc-24-v-dc-197720.html
+#
+"""
+
+import serial
+import os
+import logging
+
+from pdudaemon.drivers.driver import PDUDriver
+log = logging.getLogger("pdud.drivers." + os.path.basename(__file__))
+
+# number of retries on communication error
+RETRY_COUNT=1
+
+# timeout while receiving
+SERIAL_TIMEOUT=.1
+
+# given by hardware
+PORTS_PER_CARD=8
+MAX_CARDS=255
+
+# as described in manual, address 0 for first card does not work
+ADDR_FIRST_CARD=1
+
+# valid commands
+CMD_INIT=1
+CMD_GETPORT=2
+CMD_SETPORT=3
+CMD_GETOPTION=4
+CMD_SETOPTION=5
+CMD_SETSINGLE=6
+CMD_DELSINGLE=7
+CMD_TOGGLESINGLE=8
+
+# parameter for CMD_SETOPTION / CMD_GETOPTION
+OPTION_BROADCAST_NO_SEND_AND_NO_BLOCK=0
+OPTION_BROADCAST_SEND_AND_NO_BLOCK=1 # default
+OPTION_BROADCAST_NO_SEND_AND_BLOCK=2
+OPTION_BROADCAST_SEND_AND_BLOCK=3
+
+class Conrad197720(PDUDriver):
+    """Driver for Conrad Components 197720 and 197730 relay card
+
+    https://www.conrad.com/p/conrad-components-197720-relay-card-component-12-v-dc-24-v-dc-197720
+    https://www.conrad.com/p/conrad-components-197730-relay-card-component-12-v-dc-197730
+
+    Up to 255 relay cards can be used on one serial port
+    The difference between model 197720 and model 197730 is the switchting power of the relays
+    The protocol is the same for both card types
+    """
+    def __init__(self, hostname, settings):
+        self.com = serial.Serial()
+        self.num_cards = 0
+        self.hostname = hostname
+        self.openConnection(settings.get("device", "/dev/ttyUSB0")) # /dev/ttyUSB0 is default
+        self.init()
+
+    def __del__(self):
+        if self.com.is_open:
+            self.com.close()
+
+    @classmethod
+    def accepts(cls, drivername):
+        return drivername == "conrad197720"
+
+    def port_interaction(self, command, port_number):
+        """
+        pdudaemon method for port interaction
+        :param self: The object itself
+        :param command: The command string
+        :param port_number: The port number 0...n
+        """
+        self.updateSingle(command, port_number)
```

Comment:
port_number need to be converted to an int (see https://github.com/pdudaemon/pdudaemon/pull/135).

Reply:
I will change that, thanks for the hint!